### PR TITLE
Batching, instancing, and rendering performance stuff

### PR DESCRIPTION
**Pull request by superdump**

### CODE REVIEW

**Comment by rparrett** [comment]
It's really cool to read through the explanation of some of these PRs that were a bit over my head :)

Some feedback on this section:

> What is next for batching/instancing and beyond?

`This is implemented on a branch.` this phrasing seems potentially confusing. **I** know you mean that you or someone else has demonstrated this as a proof of concept (either privately or publicly) but I'm not sure if the blog-reader might assume this code is in the main branch, or there's an existing PR, or what.

Maybe consider splitting the bullet points into "stuff that has been demonstrated as a proof of concept, but needs work" and things that are possible but are still in the private / imagination stage, with some sort of clarifying text between?

Also, there was a great effort in previous sections to lay out the webgl2 story, and **maybe** it would be nice to mention that here as well -- which of these potential improvements are applicable there? How much are we just banking on webgpu saving us before those land? How is webgpu availability coming along, anyway? I have those sorts of questions after reading the section.

> Rob Swain (@superdump) intends to implement an alternative method similar to what is done in rend3.

Absolutely no idea what rend3 is or how it works and I suspect the average blog-reader is in a similar position.

File: content/news/2023-10-21-bevy-0.12/index.md [hunk 1]
```diff
@@ -19,6 +19,248 @@ Since our last release a few months ago we've added a _ton_ of new features, bug
...
+
+However, `SparseSet` has the downside that the dense `Vec<usize>` has to be as large as the largest contained `Entity.index`. If you spawn a million entities, then despawn 999,999, leaving the millionth entity still spawned, every `SparseSet<Entity, T>` for different `T` will have to have a `Vec<usize>` that is one million items large.
+
+`HashMap<Entity, T>` is similar to `SparseSet`, and more familiar. It has good space complexity, with performance depending a lot on the hash function.
+
+Fast hash functions from the wild were tested AHash, FNV, FxHasher, SeaHasher, but all had quite a big performance drop compared to `SparseSet`. Ultimately, a hash function designed by @SkiFire13, and inspired by `rustc-hash`, was chosen that has strong performance and is robust enough for this usage. This combination is called `EntityHashMap` and is the new way to store component data in the render world.
+
+The worst case performance with random Z-order spawning of 2D meshes/sprites in `bevymark` is similar to `SparseSet`. The best case performance with 2D meshes/sprites spawned in draw order
```
[the remark this answers is not in the document]
Comment (superdump) [reply content/news/2023-10-21-bevy-0.12/index.md:223]: Indeed it is.

**Comment by superdump** [comment]
> It's really cool to read through the explanation of some of these PRs that were a bit over my head :)

Are these descriptions easier to understand? Or is it still over your head? Does it feel like too much detail?

> Some feedback on this section:
> 
> > What is next for batching/instancing and beyond?
> 
> `This is implemented on a branch.` this phrasing seems potentially confusing. **I** know you mean that you or someone else has demonstrated this as a proof of concept (either privately or publicly) but I'm not sure if the blog-reader might assume this code is in the main branch, or there's an existing PR, or what.

I could reword that to something like "a proof of concept implementation shows promising results" or similar.

> Maybe consider splitting the bullet points into "stuff that has been demonstrated as a proof of concept, but needs work" and things that are possible but are still in the private / imagination stage, with some sort of clarifying text between?

Good idea.

> Also, there was a great effort in previous sections to lay out the webgl2 story, and **maybe** it would be nice to mention that here as well -- which of these potential improvements are applicable there? How much are we just banking on webgpu saving us before those land? How is webgpu availability coming along, anyway? I have those sorts of questions after reading the section.

I will clarify this, but noting the answer here: All of this is designed with WebGL2 support in mind and all the changes to avoid rebinding benefit WebGL2 as well. That said, WebGL2 stops at CPU-driven rendering. WebGPU and native can benefit from GPU-driven, although WebGPU does not have support for bindless texture arrays, nor does it have support for 'multi-draw indirect'. Bindless texture arrays can be 'replaced' to some degree by texture atlases in array textures. I will add a note about this. Multi-draw indirect can be replaced by many single-draw indirect, and indeed wgpu has to emulate this on some platforms (Metal) anyway. An indirect draw is one where the vertex offset/count, instance offset/count information is written to a GPU buffer (which is how it can be GPU-driven - a compute shader can write the values into a storage buffer) and then the CPU just says draw what it says at this offset in this buffer. WebGPU is already supported by the way. All I am talking about is optimisations and better leveraging what is available whether CPU-driven or enabling GPU-driven.

> > Rob Swain (@superdump) intends to implement an alternative method similar to what is done in rend3.
> 
> Absolutely no idea what rend3 is or how it works and I suspect the average blog-reader is in a similar position.

Fair point. I'll have a think about how to very very briefly express it.

**Comment by superdump** [comment]
> > Are these descriptions easier to understand? Or is it still over your head? Does it feel like too much detail?
> 
> Maybe all of the above, haha!

:)

> In particular, I feel like (haven't gone back and looked to check if my recollection matches reality) the necessity / results of reordering of the render sets was explained much better here than in that PR.

Great!

> And the case for alternative render stuff storage was laid out much more compellingly.

Images will come too, hopefully, that illustrate the rebinding frequency difference.

> I'd say that the hashmap section was the one my eyes glossed over the most. Maybe a bit too much detail there for me personally. But in general, I really appreciate laying out the case for users for these sorts of big changes so I don't really mind.

I very strongly suspected that as I wrote it. I could probably cut that one down to focus on just avoiding the table moves and this is the solution, rather than discussing all the details of how EntityHashMap was chosen. What do you think?

> The third paragraph of the sprite instancing section is still pretty unclear to me. I think it would be a good idea to define "instance-rate buffer" a bit better somewhere.

I could remove the some of that as too much information and just tl;dr it faster.

**Comment by rparrett** [comment]
> I very strongly suspected that as I wrote it. I could probably cut that one down to focus on just avoiding the table moves and this is the solution, rather than discussing all the details of how EntityHashMap was chosen. What do you think?

Hmm yeah, maybe another opinion is in order, but I think you could leave out a lot of the `Vec<T>` content and summarize it in a "future work" paragraph or something.

File: content/news/2023-10-21-bevy-0.12/index.md [hunk 2]
```diff
@@ -19,6 +19,301 @@ Since our last release a few months ago we've added a _ton_ of new features, bug
...
+
+3D Meshes
+
+2D Meshes
+
+Sprites
+
+UI
+
+### What's next for rendering performance?
```
Comment (JMS55) [inline content/news/2023-10-21-bevy-0.12/index.md:311]: Imo we should mention parallel draw encoding, which I have a PR open for. We're just waiting for wgpu to land arcanization.

**Review by JMS55 — CHANGES_REQUESTED** [review]
Overall, mostly fine? I'm just worried it's a bit too technical/disorganized. The differences split between the "general renderer perf improvements" and batching related perf improvements are a bit confusing, it feels like we're talking about the same thing twice. I feel like people will read the first section, and then get to the second and go "wait, I thought we already finished with the renderer perf improvements". Not sure what a better approach would be, and ultimately writing is hard, so /shrug.

Maybe a high level overview, just a bulleted list of perf improvements at the start, would be a good idea, before diving into specific changes.

File: content/news/2023-10-21-bevy-0.12/index.md [hunk 3]
```diff
@@ -19,6 +19,319 @@ Since our last release a few months ago we've added a _ton_ of new features, bug
...
+
+Bevy's renderer performance for 2D and 3D meshes can improve a lot. Both CPU and graphics API / GPU bottlenecks can be removed to give significantly higher frame rates. As always with Bevy, we want to make the most of the platforms you use, from the constraints of WebGL2 and mobile devices, to the highest-end native discrete graphics cards. A solid foundation can support all of this.
+
+### What are the bottlenecks?
+
+One major bottleneck is the structure of the data used for rendering.
+
+* Mesh entity data is stored in one uniform buffer, but has to be rebound at different dynamic offsets for every single draw.
+* Material type data (e.g. `StandardMaterial` uniform properties, but not textures) are stored in individual uniform buffers that have to be rebound per draw if the material changes.
+* Material textures are stored individually and have to be rebound per draw if a mesh texture changes.
```
Comment (cart) [inline content/news/2023-10-21-bevy-0.12/index.md:34]: This is an important part of optimizing / prepping for GPU-driven rendering, but it does not directly relate to changes in this release. I think we should cut it.

File: content/news/2023-10-21-bevy-0.12/index.md [hunk 4]
```diff
@@ -19,6 +19,319 @@ Since our last release a few months ago we've added a _ton_ of new features, bug
...
 
+## Automatic Batching and Instancing, and the Road to GPU-driven Rendering
+
+<div class="release-feature-authors">authors: Rob Swain (@superdump), @james-j-obrien, @JMS55, @inodentry, @robtfm, @nicopap, @teoxoy, @IceSentry, @Elabajaba</div>
+
+Bevy's renderer performance for 2D and 3D meshes can improve a lot. Both CPU and graphics API / GPU bottlenecks can be removed to give significantly higher frame rates. As always with Bevy, we want to make the most of the platforms you use, from the constraints of WebGL2 and mobile devices, to the highest-end native discrete graphics cards. A solid foundation can support all of this.
+
+### What are the bottlenecks?
+
+One major bottleneck is the structure of the data used for rendering.
```
Comment (cart) [inline content/news/2023-10-21-bevy-0.12/index.md:30]: Rather than listing all of the bottlenecks up front, I think we should let each feature section describe how it relieves a bottleneck (or prepares to relieve one).

File: content/news/2023-10-21-bevy-0.12/index.md [hunk 5]
```diff
@@ -19,6 +19,319 @@ Since our last release a few months ago we've added a _ton_ of new features, bug
...
+<div style="font-size: 1.0rem" class="release-feature-authors">Tested on an M1 Max, limiting the BatchedUniformBuffer batch size to 1 versus how it works in 0.12.</div>
+
+3D meshes: `many_cubes` which spawn 160,000 cubes, of which ~11,700 are visible in the view. These are drawn using a single instanced draw of all visible cubes which enables up to **100% increase in frame rate (2x)**!
+
+![0.12-3DMeshes](0.12-3DMeshes.svg)
+<div style="font-size: 1.0rem" class="release-feature-authors">Tested on an M1 Max, limiting the BatchedUniformBuffer batch size to 1 versus how it works in 0.12.</div>
+
+These performance benefits can be leveraged on all platforms, including WebGL2!
+
+#### What is next for batching/instancing and beyond?
```
Comment (cart) [inline content/news/2023-10-21-bevy-0.12/index.md:241]: This section is full of too much context. This should be reduced to a few short / one sentence bullet points and moved to the generic "whats next" section at the end.

File: content/news/2023-10-21-bevy-0.12/index.md [hunk 6]
```diff
@@ -19,6 +19,319 @@ Since our last release a few months ago we've added a _ton_ of new features, bug
...
+`GpuArrayBuffer` was designed and implemented as an abstraction over `BatchedUniformBuffer` and using a `StorageBuffer` to store an array of `T`.
+
+The data in a `StorageBuffer` looks like this:
+
+![StorageBuffer](StorageBuffer.svg)
+<div style="font-size: 1.0rem" class="release-feature-authors">Red arrows are 'rebinds', blue boxes are instance data.</div>
+
+All the instance data can be placed directly one after the other, and we only have to bind once. There is no need for any dynamic offset binding, so there is no need for any padding for alignment.
+
+```rust
```
Comment (cart) [inline content/news/2023-10-21-bevy-0.12/index.md:112]: This code example would be good as documentation / Bevy Book content, but I think its just slightly too technical / verbose for the average reader. It cross cuts into a lot of different features (shaders, shader defs, ShaderType/encase, GpuArrayBuffer, pipeline layouts). I think the percentage of readers that will understand whats happening here is significantly less than %1. I think we can get away with cutting it.

File: content/news/2023-10-21-bevy-0.12/index.md [hunk 7]
```diff
@@ -19,6 +19,319 @@ Since our last release a few months ago we've added a _ton_ of new features, bug
...
+Bevy's renderer performance for 2D and 3D meshes can improve a lot. Both CPU and graphics API / GPU bottlenecks can be removed to give significantly higher frame rates. As always with Bevy, we want to make the most of the platforms you use, from the constraints of WebGL2 and mobile devices, to the highest-end native discrete graphics cards. A solid foundation can support all of this.
+
+### What are the bottlenecks?
+
+One major bottleneck is the structure of the data used for rendering.
+
+* Mesh entity data is stored in one uniform buffer, but has to be rebound at different dynamic offsets for every single draw.
+* Material type data (e.g. `StandardMaterial` uniform properties, but not textures) are stored in individual uniform buffers that have to be rebound per draw if the material changes.
+* Material textures are stored individually and have to be rebound per draw if a mesh texture changes.
+* Mesh index / vertex buffers are stored individually per-mesh and have to be rebound per draw.
```
Comment (cart) [inline content/news/2023-10-21-bevy-0.12/index.md:35]: This doesn't relate to work that was done this release. I think we should cut it.

File: content/news/2023-10-21-bevy-0.12/index.md [hunk 8]
```diff
@@ -19,6 +19,319 @@ Since our last release a few months ago we've added a _ton_ of new features, bug
...
+The bevymark example needed to be improved to enable benchmarking the batching / instanced draw changes. Modes were added to:
+
+* draw 2D quad meshes instead of sprites: `--mode mesh2d`
+* vary the per-instance color data instead of only varying the colour per wave of birds: `--vary-per-instance`
+* generate a number of material / sprite textures and randomly choose from them either per wave or per instance depending on the vary per instance setting: `--material-texture-count 10`
+* spawn the birds in random z order (new default), or in draw order: `--ordered-z`
+
+This allows benchmarking of different situations for batching / instancing in the next section.
+
+### Automatic Batching/Instancing of Draw Commands
```
Comment (cart) [inline content/news/2023-10-21-bevy-0.12/index.md:199]: I think we should move this section higher up. I know it builds on other things, but it feels like we're burying the lead here. There are ways to reference an abstraction used prior to defining it.

File: content/news/2023-10-21-bevy-0.12/index.md [hunk 9]
```diff
@@ -19,6 +19,319 @@ Since our last release a few months ago we've added a _ton_ of new features, bug
 
 <div class="release-feature-authors">authors: @author</div>
 
+## Automatic Batching and Instancing, and the Road to GPU-driven Rendering
+
+<div class="release-feature-authors">authors: Rob Swain (@superdump), @james-j-obrien, @JMS55, @inodentry, @robtfm, @nicopap, @teoxoy, @IceSentry, @Elabajaba</div>
+
+Bevy's renderer performance for 2D and 3D meshes can improve a lot. Both CPU and graphics API / GPU bottlenecks can be removed to give significantly higher frame rates. As always with Bevy, we want to make the most of the platforms you use, from the constraints of WebGL2 and mobile devices, to the highest-end native discrete graphics cards. A solid foundation can support all of this.
```
Comment (alice-i-cecile) [inline content/news/2023-10-21-bevy-0.12/index.md:26]:
```suggestion
Bevy's renderer performance for 2D and 3D meshes can improve a lot. There are bottlenecks on both the CPU and GPU side, which can be lessened to give significantly higher frame rates. As always with Bevy, we want to make the most of the platforms you use, from the constraints of WebGL2 and mobile devices, to the highest-end native discrete graphics cards. A solid foundation can support all of this.
```

File: content/news/2023-10-21-bevy-0.12/index.md [hunk 10]
```diff
@@ -19,6 +19,319 @@ Since our last release a few months ago we've added a _ton_ of new features, bug
...
+One major bottleneck is the structure of the data used for rendering.
+
+* Mesh entity data is stored in one uniform buffer, but has to be rebound at different dynamic offsets for every single draw.
+* Material type data (e.g. `StandardMaterial` uniform properties, but not textures) are stored in individual uniform buffers that have to be rebound per draw if the material changes.
+* Material textures are stored individually and have to be rebound per draw if a mesh texture changes.
+* Mesh index / vertex buffers are stored individually per-mesh and have to be rebound per draw.
+
+All of this rebinding has both CPU and graphics API / GPU performance impact. On the CPU, it means encoding of draw commands has many more steps to process and so takes more time than necessary. In the graphics API and on the GPU, it means many more rebinds, and separate draw commands.
+
+Avoiding rebinding is both a big performance benefit for CPU-driven rendering, including WebGL2, and is necessary to enable GPU-driven rendering.
```
Comment (alice-i-cecile) [inline content/news/2023-10-21-bevy-0.12/index.md:39]: Needs an ELI5 about bindings (and thus rebinding) for non-graphics programmers.

File: content/news/2023-10-21-bevy-0.12/index.md [hunk 11]
```diff
@@ -19,6 +19,319 @@ Since our last release a few months ago we've added a _ton_ of new features, bug
...
+
+### What are CPU- and GPU-driven rendering?
+
+CPU-driven rendering is where draw commands are created on the CPU, in Bevy this means in Rust code, more specifically in render graph nodes.
```
Comment (alice-i-cecile) [inline content/news/2023-10-21-bevy-0.12/index.md:43]:
```suggestion
CPU-driven rendering is where draw commands are created on the CPU. In Bevy this means in Rust code, more specifically in render graph nodes.
```

File: content/news/2023-10-21-bevy-0.12/index.md [hunk 12]
```diff
@@ -19,6 +19,319 @@ Since our last release a few months ago we've added a _ton_ of new features, bug
...
+
+In GPU-driven rendering, the draw commands are encoded on the GPU by compute shaders. This leverages GPU parallelism, and unlocks more advanced culling optimisations that are infeasible to do on the CPU, among many other methods that bring large performance benefits.
```
Comment (alice-i-cecile) [inline content/news/2023-10-21-bevy-0.12/index.md:45]:
```suggestion
In GPU-driven rendering, the draw commands are encoded on the GPU by [compute shaders](https://www.khronos.org/opengl/wiki/Compute_Shader). This leverages GPU parallelism, and unlocks more advanced culling optimizations that are infeasible to do on the CPU, among many other methods that bring large performance benefits.
```

File: content/news/2023-10-21-bevy-0.12/index.md [hunk 13]
```diff
@@ -19,6 +19,319 @@ Since our last release a few months ago we've added a _ton_ of new features, bug
...
+
+![RenderSets-0.12](RenderSets-0.12.svg)
+
+`PrepareAssets` was introduced because we only want to queue entities for drawing if their assets have been prepared. Per-frame data preparation still happens in the `Prepare` set, specifically in its `PrepareResources` subset. That is now after `Queue` and `Sort`, so the order of draws is known. This also made a lot more sense for batching, as it is now known at the point of batching whether an entity that is of another type in the render phase needs to be drawn. Bind groups now have a clear subset where they should be created - `PrepareBindGroups`.
+
+### BatchedUniformBuffer and GpuArrayBuffer
+
+OK, so we need to put many pieces of data of the same type into buffers in a way that we can bind them as few times as possible and draw multiple instances from them. How can we do that?
+
+In 0.11 per-instance `MeshUniform` data is stored in a uniform buffer with each instance's data aligned to a dynamic offset. When drawing each mesh entity, we update the dynamic offset, which is close to rebinding. It looks like this:
```
Comment (alice-i-cecile) [inline content/news/2023-10-21-bevy-0.12/index.md:71]: I don't understand what "which is close to rebinding" means

File: content/news/2023-10-21-bevy-0.12/index.md [hunk 14]
```diff
@@ -19,6 +19,319 @@ Since our last release a few months ago we've added a _ton_ of new features, bug
...
+![DynamicUniformBuffer](DynamicUniformBuffer.svg)
+<div style="font-size: 1.0rem" class="release-feature-authors">Red arrows are 'rebinds' to update the dynamic offset, blue boxes are instance data, orange boxes are padding for dynamic offset alignment, which is a requirement of GPUs and graphics APIs.</div>
+
+Instance-rate vertex buffers are one way, but they are very constrained to having a specific order. They are/may be suitable for per-instance data like mesh entity transforms, but they can't be used for material data. The other main options are uniform buffers, storage buffers, and data textures.
+
+WebGL2 does not support storage buffers, only uniform buffers. Uniform buffers have a minimum guaranteed size per binding of 16kB on WebGL2. Storage buffers, where available, have a minimum guaranteed size of 128MB.
+
+Data textures are far more awkward for structured data, and without support for linear data layouts on some platforms, they will perform worse.
+
+We want to support uniform buffers on WebGL2 or where storage buffers are not supported, and use storage buffers everywhere else.
```
Comment (alice-i-cecile) [inline content/news/2023-10-21-bevy-0.12/index.md:82]:
```suggestion
We want to support uniform buffers where storage buffers are not supported (like WebGL2) and use storage buffers everywhere else.
```

File: content/news/2023-10-21-bevy-0.12/index.md (hunk 6 above)
Comment (alice-i-cecile) [reply content/news/2023-10-21-bevy-0.12/index.md:112]: I would cut this. Too long and crunchy I think.

File: content/news/2023-10-21-bevy-0.12/index.md [hunk 15]
```diff
@@ -19,6 +19,319 @@ Since our last release a few months ago we've added a _ton_ of new features, bug
...
+
+Part of the design involves clearing the render world of all entities between frames. This enables consistent Entity mapping between the main and render worlds while still being able to spawn new entities in the render world that don't exist in the main world.
+
+Unfortunately, this ECS usage pattern also incurred some significant performance problems. Entities are cleared and respawned each frame, components are inserted across many systems and different parts of the render app schedule.
+
+The fastest ECS storage available is called table storage. A simplified concept for table storage is that it is a structure of arrays of component data. Each archetype has its own table for storage. Whenever a new component is inserted onto an entity that it didn't have before, its archetype is changed. This then requires that that entity's component data be moved from the table for the old archetype to the table for the new archetype.
+
+In practice this was very visible in profiles as long-running system commands throughout the render app schedule.
+
+DEMO PROFILE IMAGE
```
Comment (alice-i-cecile) [inline content/news/2023-10-21-bevy-0.12/index.md:271]: This appears to be missing.

File: content/news/2023-10-21-bevy-0.12/index.md [hunk 16]
```diff
@@ -19,6 +19,319 @@ Since our last release a few months ago we've added a _ton_ of new features, bug
...
+
+3D Meshes
+
+2D Meshes
+
+Sprites
+
+UI
+
+### What's next for rendering performance?
```
Comment (alice-i-cecile) [inline content/news/2023-10-21-bevy-0.12/index.md:329]: This should also be moved into "What's Next" at the bottom of the post.

File: content/news/2023-10-21-bevy-0.12/index.md (hunk 13 above)
Comment (superdump) [reply content/news/2023-10-21-bevy-0.12/index.md:71]: A buffer binding is telling the graphics API that that buffer, at that memory address, and with that size needs to be accessible for the duration of execution of subsequent draws until the binding is changed. A dynamic offset is just a byte offset to the memory address of the start of the binding. In that sense, updating the dynamic offset, while a bit cheaper through some graphics APIs/drivers than updating the entire binding, is close to just updating that entire binding.

File: content/news/2023-10-21-bevy-0.12/index.md (hunk 15 above)
Comment (superdump) [reply content/news/2023-10-21-bevy-0.12/index.md:271]: Yup. It is.

**Comment by cart** [comment]
Just pushed two commits with the goals of:

1. Slimming down, especially on the technical details and "details about future work".
2. Moving relevant bits to the "Whats Next" section.
3. Reordering by "precedence / interest"
4. Adding clarity + easing people into the details
5. Adding documentation

**Comment by cart** [comment]
I also removed the "total 0.11 vs 0.12 performance difference" section and other placeholder content. We can follow up later with those if we find time.

If I find time I also might re-do the graphs to use frame time instead of FPS (and to include values in the graphs). But for now I think this should be mergeable.

File: content/news/2023-10-21-bevy-0.12/index.md [hunk 17]
```diff
@@ -19,56 +19,100 @@ Since our last release a few months ago we've added a _ton_ of new features, bug
...
-Bevy's renderer performance for 2D and 3D meshes can improve a lot. There are bottlenecks on both the CPU and GPU side, which can be lessened to give significantly higher frame rates. As always with Bevy, we want to make the most of the platforms you use, from the constraints of WebGL2 and mobile devices, to the highest-end native discrete graphics cards. A solid foundation can support all of this.
+**Bevy 0.12** now automatically batches/instances draw commands where possible. This cuts down the number of draw calls, which yields significant performance wins!
+
+This required a number of architectural changes, including how we store and access per-entity mesh data (more on this later).
+
+Here are some benches of the old unbatched approach (0.11) to the new batched approach (0.12):
+
+### 2D Mesh Bevymark (frames per second, more is better)
+
+This renders 160,000 entities with textured quad meshes (160 groups of 1,000 entities each, each group sharing a material). This means we can batch each group, resulting in only 160 instanced draw calls when batching is enabled.
```
Comment (superdump) [inline content/news/2023-10-21-bevy-0.12/index.md:34]: This is misleading - as the batch size for the graph below was set to 1, it will do 160k draws I think.

File: content/news/2023-10-21-bevy-0.12/index.md [hunk 18]
```diff
@@ -98,178 +142,60 @@ Notice how the instance data can be packed much more tightly, fitting the same a
...
-
-### EntityHashMap
+## EntityHashMap Renderer Optimization
 
 <div class="release-feature-authors">authors: Rob Swain (@superdump), @robtfm, @pcwalton, @jancespivo, @SkiFire13, @nicopap</div>
 
-#### The Performance Problem
-
-Since Bevy 0.6, Bevy's renderer has used a separate render world to store an extracted snapshot of the simulated data from the main world to enable pipelined rendering of frame N in the render app, while the main app simulates frame N+1.
+Since **Bevy 0.6**, Bevy's renderer has extracted data from the "app world" into a separate "render world". This enables [Pipelined Rendering](/news/bevy-0-6/#pipelined-rendering-extract-prepare-queue-render), which renders frame N in the render app, while the main app simulates frame N+1.
```
Comment (superdump) [inline content/news/2023-10-21-bevy-0.12/index.md:173]: @cart as far as I am aware we always call them the main world and render world. Or main app and render app, depending on whether we’re talking about the world or app specifically.

Comment (cart) [reply content/news/2023-10-21-bevy-0.12/index.md:173]: Yup thats preferable!

File: content/news/2023-10-21-bevy-0.12/index.md (hunk 13 above)
Comment (cart) [reply content/news/2023-10-21-bevy-0.12/index.md]: I just clarified this with "close in cost"

File: content/news/2023-10-21-bevy-0.12/index.md (hunk 2 above)
Comment (cart) [reply content/news/2023-10-21-bevy-0.12/index.md]: I've slimmed down these "whats next" sections a lot. I think we should just cover this whenever we've finished it (or when we've kicked off the effort).
